osmo_jupyter/plot.py

```diff
@@ -0,0 +1,277 @@
+    * The primary Y axis (y0) is reserved for annotations. This is done to guarantee that annotations show
+    * If colors are on separate axes, they will be on the next consecutive axes, otherwise they will be on yaxis 2
+    * You can use additional axes for plotting other things by passing in additional params
+
+    Args:
+        x_axis_title: title to use for X axis, default 'Time'
+        y_axis_title: title to use for Y axis where colors are plotted. Required if colors_on_separate_axes is False.
+            If colors_on_separate_axes is True, it will be used as a suffix for the y axis titles "r", "g", and "b"
+        colors_on_separate_axes: whether colors (r,g and b) should be displayed on separate Y axes
+        events: dictionary of {"annotation": x_axis_value} which will be used to annotate the chart
+        additional_layout_kwargs: Any additional arguments will be passed to go.Layout. These can, for instance,
+            be the result of _axis_kwargs() to add support for graphing more items.
+            Valid options are documented at https://plot.ly/python/reference/#layout
+    '''
+    events = events or {}
+
+    if not colors_on_separate_axes and not y_axis_title:
+        raise ValueError('If colors_on_separate_axes is False you must provide a y_axis_title')
+
+    def get_color_y_axis_title(color):
+        if not colors_on_separate_axes:
+            return y_axis_title
+
+        # blue and green are both on the right side, and applying a line break to blue lets them both be there
+        # without overlapping
+        line_break = '<br>' if color is 'b' else ''
+
+        # If colors are on separate axes and a Y axis title is provided, use it as a suffix
+        suffix = f' {y_axis_title}' if y_axis_title else ''
+        return f'{line_break}{color}{suffix}'
+
+    colors_on_separate_axes_kwargs = _add_dicts(
+        _axis_kwargs(
+            axis_number=AXES_BY_COLOR[color],
+            color=COLORS_BY_LETTER[color],
+            title=get_color_y_axis_title(color),
+            overlaying='y',
+            side=AXIS_SIDES_BY_COLOR[color],
+        )
+        for color in COLOR_CHANNELS
+    )
+    colors_on_same_axes_kwargs = _axis_kwargs(
+        axis_number=SHARED_COLOR_AXIS,
+        title=y_axis_title,
+        overlaying='y',
+        side='left',
+    )
+    color_y_axes = colors_on_separate_axes_kwargs if colors_on_separate_axes else colors_on_same_axes_kwargs
+
+    # Prepare annotations & annotation axis
+
+    # By default, the primary Y axis is invisible and used for annotations only,
+    # but note that it can be overridden using additional_layout_kwargs
+    primary_y_axis = {
+        'yaxis': {
+            'title': '',
+            'visible': False,
+        }
+    }
+    # We need to have at least 1 annotation on axis 1 or else plotted values will disappear
+    # https://github.com/plotly/plotly.py/issues/1200
+    dummy_annotation = [{
+        'yref': 'y',
+        'xref': 'paper',
```

Comment:
what is 'paper'?

Reply:
it references the location of the annotation to the "canvas", not to the axis values. so 0.5 would be the middle no matter what the X values are.
I'm still fiddling with this a bit since the multiple-y-axes bug has resurfaced :( :(

Reply:
I (mostly) figured it out. They key is how the `overlaying` argument works, and I think I mostly know what it means now.